### ISSUE TITLE
Flesh out the "common patterns" section of the PEP

### ISSUE THREAD

**Issue by davepeck** [davepeck/peps#4]
Include:

- [x] A section on parsing to an AST
- [x] A section about context-sensitive interpolation 
- [x] A section about nested template strings

**Comment by ncoghlan** [comment]
While it's more a "possible" pattern than a "common" pattern, the `on_format` lazy evaluation idea from the last discussion may be worth mentioning: https://discuss.python.org/t/pep-750-tag-strings-for-writing-domain-specific-languages/60408/199

```
class on_format:
    def __init__(self, callable):
        self._callable = callable

    def __format__(self, spec):
        return format(self._callable(), spec)
```
```
do_something_with(t'{on_format(lambda: 42)}')```
```

Using `__format__` specifically this way came up in the context of wanting a way to delay code execution for expensive calls only used for logging purposes. Other template renderers may prefer to use more specific methods, so the general pattern is just about having a known method call that can be intercepted to support opt-in lazy evaluation.

**Comment by davepeck** [comment]
I took a stab at this. It will definitely need some word-smithing/clarification but it's a start. Closed with 556b316f
